websockets/cext.c

```diff
@@ -0,0 +1,128 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+
+#include <stdlib.h> /* getenv */
+
+#ifndef HAS_SSE2
+/* setup HAS_SSE2 macro according to compiler predefined macros */
+#ifdef __SSE2__
+#define HAS_SSE2 1
+#else
+#define HAS_SSE2 0
+#endif
+#endif /* not defined HAS_SSE2 */
+
+#if HAS_SSE2
+#include <emmintrin.h>
+#endif
+
+#ifndef HAS_FAST_UNALIGNED_ACCESS
+/* This is not enabled by default */
+/* build with CFLAGS="-DHAS_FAST_UNALIGNED_ACCESS=1" to enable */
+/* There seem to be no information regarding alignment of PyBytes allocated buffer */
+#define HAS_FAST_UNALIGNED_ACCESS 0
+#endif /* not defined HAS_FAST_UNALIGNED_ACCESS */
+
+#if HAS_FAST_UNALIGNED_ACCESS
+#include <stdint.h>
+#endif
+
+static PyObject* apply_mask(PyObject* self, PyObject* args) {
+    const char* mask;
+    const char* in;
+    char* out;
+    Py_ssize_t mask_len;
+    Py_ssize_t len;
+    Py_ssize_t i = 0; /* initialize iterator */
+    PyObject* result;
+
+
+    if (!PyArg_ParseTuple(args, "s#s#", &mask, &mask_len, &in, &len)) {
+        return NULL;
+    }
+
+    result = PyBytes_FromStringAndSize(NULL, len);
```

Comment:
In branches that vectorize or unroll the loop, I think it would be more readable to allocate a slightly larger buffer, with a length that's a multiple of 16 bytes (for your SSE2 implementation) or 8 bytes / 4 bytes (for casting to 64 bits / 32 bits integers and XORing integers, which should be even faster than XORing bytes, unless the compiler optimizes that already).

Then, after running the loop, we can [resize](https://docs.python.org/3/c-api/bytes.html#c._PyBytes_Resize) the buffer to the target size, which will happen in place when the message is at least  because the object only shrinks by at most 15 / 7 / 3 bytes respectively.

For small messages (let's say up to 64 bytes) I don't expect we can do much better than the completely naive version (`for (; i < len; i++) { out[i] = in[i] ^ mask[i & 3]; }`).

I'd like to play a bit with all this to see how it works.